### ISSUE TITLE
GDScript 2.0: Zero vector when using `const` instead of `var` in a specific scenario

### ISSUE THREAD

**Issue by Ohems** [godotengine/godot#59545]
### Godot version

4.0.alpha5.official (d7d528c15)

### System information

Windows 10 Pro 21H1

### Issue description

In a very specific scenario, using `const` instead of `var` in GDScript changes the outcome, resulting in a created vector incorrectly being a zero vector. Using `const` instead of `var` should only affect the mutability of the variable, not the result of the operation.

### Steps to reproduce

The following code should work in any attached script. I have tested it in the `_ready()` function of two unrelated scene objects.

## Incorrect, with const
```
const COUNT = 10
const DIV = 5
for a in range(ceil(COUNT / DIV)):
	var vec = Vector3i(a, a, a)
	print(vec, " ", a, " ", COUNT, " ", DIV)
```
incorrectly prints
```
(0, 0, 0) 0 10 5
(0, 0, 0) 1 10 5
```
## Correct, with var
```
const COUNT = 10
var DIV = 5
for a in range(ceil(COUNT / DIV)):
	var vec = Vector3i(a, a, a)
	print(vec, " ", a, " ", COUNT, " ", DIV)
```
correctly prints
```
(0, 0, 0) 0 10 5
(1, 1, 1) 1 10 5
```

**Comment by sent44** [comment]
I do some tests and I suspect how `range()` create array, then how `for` access those array and then how engine pass those value to int parameter of c++ function.

```GDScript
for a in range(3.0):
	print(Rect2i(a, a as int, a, a))
	print(int(a), ", " , int(a as int), ", ", a, ", ", a as int)
	print(Vector2i(a, a as int))
	var i := Image.new()
	i.create(a + 1, a as int + 1, false, Image.FORMAT_R8)
	print(i.get_size() - Vector2(1, 1))
```
```
[P: (0, 0), S: (0, 0)]
0, 0, 0, 0
(0, 0)
(0, 0)
[P: (0, 1), S: (0, 0)]
4607182418800017408, 1, 1, 1
(0, 1)
(0, 1)
[P: (0, 2), S: (0, 0)]
4611686018427387904, 2, 2, 2
(0, 2)
(0, 2)
```

**Comment by Ohems** [comment]
Very good analysis! It apparently also has something to do with how constant values are evaluated since
```
var A = 3.0
for a in range(A):
	print(Rect2i(a, a as int, a, a))
	print(int(a), ", " , int(a as int), ", ", a, ", ", a as int)
	print(Vector2i(a, a as int))
	var i := Image.new()
	i.create(a + 1, a as int + 1, false, Image.FORMAT_R8)
	print(i.get_size() - Vector2(1, 1))
```
prints
```
[P: (0, 0), S: (0, 0)]
0, 0, 0, 0
(0, 0)
(0, 0)
[P: (1, 1), S: (1, 1)]
1, 1, 1, 1
(1, 1)
(1, 1)
[P: (2, 2), S: (2, 2)]
2, 2, 2, 2
(2, 2)
(2, 2)
```

**Comment by Ohems** [comment]
It seems that a valid workaround is to cast the contents of the range function to `int`, especially since I'm using `ceil` to round the number. The example works correctly with
```
for a in range(3.0 as int):
```

**Comment by sent44** [comment]
You can also do without `range()` and it is also workaround for 0 to n-1 loop
```GDScript
for i in nun:
```

also from my first test (forgot to include)
```GDScript
for a in range(3.0):
   print(typeof(a) == TYPE_FLOAT) # true
print(typeof(range(3.0)[1]) == TYPE_INT) # true
```

**Comment by dalexeev** [comment]
Looks fixed by #68125. Please check if this is true.

![](https://user-images.githubusercontent.com/47700418/218272537-8702b274-4e6e-44d9-baa9-72567def1b9e.png)

> ```gdscript
> for a in range(3.0):
>    print(typeof(a) == TYPE_FLOAT) # true
> print(typeof(range(3.0)[1]) == TYPE_INT) # true
> ```

![](https://user-images.githubusercontent.com/47700418/218272746-9a11096e-7970-4888-aa60-b18d0b2c3940.png)

![](https://user-images.githubusercontent.com/47700418/218272715-602b98df-818e-4b0c-bc4c-2d9646c286cd.png)
